tests/cpp_unit_tests/test_job_dispatch.cpp

```diff
@@ -0,0 +1,367 @@
+            Idx const n_scenarios = 7; // arbitrary non-zero value
+            auto const update_data = MockUpdateDataset(has_data, n_scenarios);
+            adapter.reset_counters();
+            auto const actual_result =
+                JobDispatch::batch_calculation(adapter, result_data, update_data, JobDispatch::sequential);
+            CHECK(expected_result == actual_result);
+            // n_scenarios calculations should be done as we run sequentially
+            CHECK(adapter.get_calculate_counter() == n_scenarios);
+            CHECK(adapter.get_cache_calculate_counter() == 1); // cache calculation is done
+        }
+    }
+    SUBCASE("Test single_thread_job") {
+        auto counter = std::make_shared<CallCounter>();
+        auto adapter = JobAdapterMock{counter};
+        auto result_data = MockResultDataset{};
+        bool const has_data{false};
+        Idx const n_scenarios = 9; // arbitrary non-zero value
+        auto const update_data = MockUpdateDataset(has_data, n_scenarios);
+        exceptions.resize(n_scenarios);
+        Idx start{};
+        Idx stride{};
+        Idx call_number{};
+
+        auto get_call_number = [](Idx start, Idx stride, Idx n_scenarios) {
+            REQUIRE_MESSAGE(
+                stride > 0,
+                "Can't have stride of (less than) zero; this should be caught by a different job dispatch handling");
+            return (n_scenarios - start + stride - 1) / stride;
+        };
+
+        auto check_call_numbers = [](JobAdapterMock& adapter, Idx expected_calls) {
+            CHECK(adapter.get_setup_counter() == expected_calls);
+            CHECK(adapter.get_winddown_counter() == expected_calls);
+            CHECK(adapter.get_calculate_counter() == expected_calls);
+            CHECK(adapter.get_thread_safe_add_calculation_info_counter() == 1); // always called once
+        };
+
+        adapter.prepare_job_dispatch(update_data); // replicate preparation step from batch_calculation
+        auto single_job = JobDispatch::single_thread_job(adapter, result_data, update_data, exceptions);
+
+        adapter.reset_counters();
+        start = 0;
+        stride = 1;
+        call_number = get_call_number(start, stride, n_scenarios);
+        CAPTURE(call_number);
+        CHECK_NOTHROW(single_job(start, stride, n_scenarios));
+        check_call_numbers(adapter, call_number);
+
+        adapter.reset_counters();
+        start = 0;
+        stride = 4;
+        call_number = get_call_number(start, stride, n_scenarios);
+        CAPTURE(call_number);
+        CHECK_NOTHROW(single_job(start, stride, n_scenarios));
+        check_call_numbers(adapter, call_number);
+
+        adapter.reset_counters();
+        start = 3;
+        stride = 2;
+        call_number = get_call_number(start, stride, n_scenarios);
+        CAPTURE(call_number);
+        CHECK_NOTHROW(single_job(start, stride, n_scenarios));
+        check_call_numbers(adapter, call_number);
```

Comment:
It seems like the `calls` can be made into Idx call_count. Its more readable that way than emplace_back(0) and check for size.